vignettes/data.Rmd

```diff
@@ -0,0 +1,186 @@
+---
+title: "dust: Comparing models and data"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{dust: Comparing models and data}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+lang_output <- function(x, lang) {
+  cat(c(sprintf("```%s", lang), x, "```"), sep = "\n")
+}
+cc_output <- function(x) lang_output(x, "cc")
+r_output <- function(x) lang_output(x, "r")
+plain_output <- function(x) lang_output(x, "plain")
+```
+
+One of our aims with `dust` was to enable the creation of fast particle filters. Most of the high level interface for this is within the package [`mcstate`](https://mrc-ide.github.io/mcstate/).  Typically, when using a `dust` model with `mcstate`, you would define a function in R which compares the model state at some point in time to some data, returning a likelihood.
+
+It is possible to implement this comparison directly in the dust model, which may slightly speed up the particle filter (because the compare function will be evaluated in parallel, and because of slightly reduced data copying), but also allows running the particle filter on a GPU (see `vignette("cuda")`).
+
+This vignette outlines the steps in implementing the comparison directly as part of the model.  This is not required for basic use of dust models, and we would typically recommend this only after your model has stabilised and you are looking to extract potential additional speed-ups or accelerate the model on a GPU.
+
+We start with a simple example, a model of volatility
+
+```{r}
+volatility <- dust::dust_example("volatility")
+```
+
+To demonstrate the approach, we simulate some data from the model itself:
+
+```{r volatility_data_create}
+data <- local({
+  mod <- volatility$new(list(alpha = 0.91, sigma = 1), 0, 1L, seed = 1L)
+  mod$set_state(matrix(rnorm(1L, 0, 1L), 1))
+  steps <- seq(0, 100, by = 1)
+  res <- mod$simulate(steps)
+  observed <- res[1, 1, -1] + rnorm(length(steps) - 1, 0, 1)
+  data.frame(step = steps[-1], observed = observed)
+})
+head(data)
+plot(observed ~ step, data, type = "o", pch = 19, las = 1)
+```
+
+As in the [`mcstate` vignette](https://mrc-ide.github.io/mcstate/articles/kalman.html) we some way of comparing model output to data.  The function used there is:
+
+```{r}
+volatility_compare <- function(state, observed, pars) {
+  dnorm(observed$observed, pars$gamma * drop(state), pars$tau, log = TRUE)
+}
+```
+
+i.e., the probability is normally distributed with mean of the equal to `gamma` multiplied by the modelled value, standard deviation of `tau` and evaluated at the observed value.  Our aim here is to adapt this so that it is implemented as part of the C++ model. This requires:
+
+* describing the data (here it will be a single numeric value, but the data could be arbitrarily complicated)
```

Comment:
```suggestion
* describing the types of the observed data at each time point in a C++ struct (here it will be a single numeric value, but the data could be arbitrarily complicated)
```